File: python/react/react.py
class InputCell:
    def __init__(self, initial_value):
        self._value = initial_value
        self.outputs = []

    def set_output(self, output_cell):
        if output_cell is not None:
            self.outputs.append(output_cell)

    @property
    def value(self):
        return self._value

    @value.setter
    def value(self, value):
        self._value = value
        if self.outputs:
            for output in self.outputs:
                output.update()


class ComputeCell:
    def __init__(self, inputs, compute_function):
        self.inputs = inputs
        for i in self.inputs:
            i.set_output(self)
        self.outputs = []
        self.callbacks = []
        self.total_updated_inputs = 0
        self.compute_function = compute_function
        self._value = self.compute_function([i.value for i in inputs])

    @property
    def value(self):
        return self._value

    @value.setter
    def value(self, value):
        self._value = value
        if self.outputs:
            for output in self.outputs:
                output.update()

    def update(self):
        old_value = self.value
        self.total_updated_inputs += 1
        if self.total_updated_inputs == len(self.inputs):
            self.value = self.compute_function([i.value for i in self.inputs])
            if self.callbacks and self.value != old_value:
                for cb in self.callbacks:
                    cb(self.value)
            self.total_updated_inputs = 0

    def set_output(self, output_cell):
        self.outputs.append(output_cell)

    def add_callback(self, callback):
        self.callbacks.append(callback)

    def remove_callback(self, callback):
        try:
            self.callbacks.remove(callback)
        except ValueError:
            pass

File: kotlin/react/src/main/kotlin/React.kt
import java.util.LinkedList

class Reactor<T> {
    // Your compute cell's addCallback method must return an object
    // that implements the Subscription interface.
    interface Subscription {
        fun cancel()
    }

    fun updateGraph(start: Cell) {
        val queue = LinkedList<Cell>()
        val visited = mutableListOf<Cell>()
        queue.add(start)
        visited.add(start)

        while(queue.isNotEmpty()) {
            val cell = queue.removeFirst()
            cell.outputs.forEach{
                if (it is ComputeCell) {
                    it.updateValue()
                }
                if (!visited.contains(it)) {
                    visited.add(it)
                    queue.add(it)
                }
            }
        }
    }

    open inner class Cell(
        v: T,
        val inputs: MutableList<Cell> = mutableListOf(),
        val outputs: MutableList<Cell> = mutableListOf()) {

        var value = v
            set(value) {
                field = value
                updateGraph(this)
            }
    }

    inner class InputCell(v: T): Cell(v)

    inner class ComputeCell(
        vararg inputCells: Cell,
        var computeFunction: (values: List<T>) -> T)
        : Cell(
            v = computeFunction(inputCells.map { it.value }),
            inputs = inputCells.toMutableList()
        ) {

        private var count = 0

        private var callbacks = mutableListOf<(values: T) -> Boolean>()

        init{
            this.inputs.forEach { it.outputs.add(this) }
        }

        fun addCallback(callback: (values: T) -> Boolean): Subscription {
            callbacks.add(callback)
            return object : Subscription {
                override fun cancel() {
                    callbacks.remove(callback)
                }
            }
        }

        fun updateValue() {
            count += 1
            if (count == this.inputs.size) {
                val newValue = this.computeFunction(this.inputs.map { it.value })
                if (newValue != this.value) {
                    this.value = newValue
                    callbacks.forEach { it(this.value) }
                }
                count = 0
            }
        }
    }
}
